java/ql/src/semmle/code/java/security/OgnlInjection.qll

```diff
@@ -0,0 +1,130 @@
+/** Provides classes to reason about OGNL injection vulnerabilities. */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.ExternalFlow
+
+/**
+ * A data flow sink for unvalidated user input that is used in OGNL EL evaluation.
+ *
+ * Extend this class to add your own OGNL injection sinks.
+ */
+abstract class OgnlInjectionSink extends DataFlow::Node { }
+
+/**
+ * A unit class for adding additional taint steps.
+ *
+ * Extend this class to add additional taint steps that should apply to the `OgnlInjectionFlowConfig`.
+ */
+class OgnlInjectionAdditionalTaintStep extends Unit {
+  /**
+   * Holds if the step from `node1` to `node2` should be considered a taint
+   * step for OGNL injection taint configurations.
+   */
+  abstract predicate step(DataFlow::Node node1, DataFlow::Node node2);
+}
+
+private class DefaultOgnlInjectionSinkModel extends SinkModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "org.apache.commons.ognl;Ognl;false;getValue;;;Argument[-1..0];ognl-injection",
+        "org.apache.commons.ognl;Ognl;false;setValue;;;Argument[-1..0];ognl-injection",
+        "org.apache.commons.ognl;Node;false;getValue;;;Argument[-1..0];ognl-injection",
+        "org.apache.commons.ognl;Node;false;setValue;;;Argument[-1..0];ognl-injection",
+        "org.apache.commons.ognl.enhance;ExpressionAccessor;true;get;;;Argument[-1];ognl-injection",
+        "org.apache.commons.ognl.enhance;ExpressionAccessor;true;set;;;Argument[-1];ognl-injection",
+        "ognl;Ognl;false;getValue;;;Argument[-1..0];ognl-injection",
+        "ognl;Ognl;false;setValue;;;Argument[-1..0];ognl-injection",
+        "ognl;Node;false;getValue;;;Argument[-1..0];ognl-injection",
+        "ognl;Node;false;setValue;;;Argument[-1..0];ognl-injection",
+        "ognl.enhance;ExpressionAccessor;true;get;;;Argument[-1];ognl-injection",
+        "ognl.enhance;ExpressionAccessor;true;set;;;Argument[-1];ognl-injection",
+        "com.opensymphony.xwork2.ognl;OgnlUtil;false;getValue;;;Argument[-1..0];ognl-injection",
+        "com.opensymphony.xwork2.ognl;OgnlUtil;false;setValue;;;Argument[-1..0];ognl-injection",
+        "com.opensymphony.xwork2.ognl;OgnlUtil;false;callMethod;;;Argument[-1..0];ognl-injection"
+      ]
+  }
+}
+
+private class DefaultOgnlInjectionSink extends OgnlInjectionSink {
+  DefaultOgnlInjectionSink() { sinkNode(this, "ognl-injection") }
+}
+
+/** The class `org.apache.commons.ognl.Ognl` or `ognl.Ognl`. */
+private class TypeOgnl extends Class {
+  TypeOgnl() {
+    this.hasQualifiedName("org.apache.commons.ognl", "Ognl") or
+    this.hasQualifiedName("ognl", "Ognl")
+  }
+}
+
+/** The class `org.apache.commons.ognl.Node` or `ognl.Node`. */
+private class TypeNode extends Interface {
+  TypeNode() {
+    this.hasQualifiedName("org.apache.commons.ognl", "Node") or
+    this.hasQualifiedName("ognl", "Node")
+  }
+}
+
+private class TypeExpressionAccessor extends Interface {
```

Comment:
QlDoc?

Reply:
Not mandatory since it's a private class, but good to have, thanks :)

Added in 8d68200.